File: raspberry/SmartHome/app/src/main/java/raspberry/smarthome/database/store/firebase/FirebaseLinkedAccountsStorage.java
package raspberry.smarthome.database.store.firebase;

import android.support.annotation.NonNull;
import android.util.Log;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import raspberry.smarthome.database.Constants;
import raspberry.smarthome.database.model.LinkedAccounts;
import raspberry.smarthome.database.store.LinkedAccountsStorage;
import raspberry.smarthome.database.store.listeners.LinkedAccountsListener;

import static raspberry.smarthome.MainActivity.DEBUG;

public class FirebaseLinkedAccountsStorage implements LinkedAccountsStorage {

    private static FirebaseLinkedAccountsStorage instance;

    private DatabaseReference ref;


    public static FirebaseLinkedAccountsStorage getInstance() {
        if(instance == null)
            instance = instantiate();

        return instance;
    }

    private static FirebaseLinkedAccountsStorage instantiate() {
        FirebaseAuth auth = FirebaseAuth.getInstance();

        if(auth.getCurrentUser() == null)
            return null;

        else return new FirebaseLinkedAccountsStorage(auth.getCurrentUser().getUid());
    }

    private FirebaseLinkedAccountsStorage(String uid) {
        FirebaseDatabase database = FirebaseDatabase.getInstance();

        ref = database.getReference()
                .child(uid)
                .child(Constants.LINKED_ACCS_REF);
    }

    @Override
    public void postLinkedAccounts(LinkedAccounts linkedAccounts) {
        ref.setValue(linkedAccounts);
    }

    @Override
    public void postLinkedAccounts(LinkedAccounts linkedAccounts, DatabaseReference.CompletionListener listener) {
        ref.setValue(linkedAccounts, listener);
    }

    @Override
    public void getLinkedAccounts(final LinkedAccountsListener listener) {
        ref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                listener.onLinkedAccountsReceived(dataSnapshot.getValue(LinkedAccounts.class));
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                if(DEBUG) Log.d(instance.getClass().getName(), Constants.FIREBASE_READ_VALUE_ERROR, databaseError.toException());
            }
        });
    }
}

File: smarthomelibary/DataLibrary/datalibrary/src/main/java/smarthome/datalibrary/database/store/firebase/FirebaseLinkedAccountsStorage.kt
package smarthome.datalibrary.database.store.firebase

import android.util.Log
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import smarthome.datalibrary.database.constants.Constants.FIREBASE_READ_VALUE_ERROR
import smarthome.datalibrary.database.constants.Constants.LINKED_ACCS_REF
import smarthome.datalibrary.database.model.LinkedAccounts
import smarthome.datalibrary.database.store.LinkedAccountsStorage
import smarthome.datalibrary.database.store.listeners.LinkedAccountsListener

class FirebaseLinkedAccountsStorage(private val uid: String,
                                    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
) : LinkedAccountsStorage {

    private val ref: DocumentReference = db.collection(uid).document(LINKED_ACCS_REF)

    override fun postLinkedAccounts(
        linkedAccounts: LinkedAccounts,
        successListener: OnSuccessListener<Void>,
        failureListener: OnFailureListener
    ) {
        ref.set(linkedAccounts)
            .addOnSuccessListener(successListener)
            .addOnFailureListener(failureListener)
    }

    override fun getLinkedAccounts(listener: LinkedAccountsListener) {
        ref.get()
            .addOnSuccessListener { res ->
                res.toObject(LinkedAccounts::class.java)?.let { listener.onLinkedAccountsReceived(it) }
            }
            .addOnFailureListener { exception -> Log.d(javaClass.name, FIREBASE_READ_VALUE_ERROR, exception) }
    }

    companion object {

        private var instance: FirebaseLinkedAccountsStorage? = null


        fun getInstance(): FirebaseLinkedAccountsStorage? {
            if (instance == null)
                instance = instantiate()

            return instance
        }

        private fun instantiate(): FirebaseLinkedAccountsStorage? {
            val auth = FirebaseAuth.getInstance()

            auth.currentUser?.let { return FirebaseLinkedAccountsStorage(it.uid) } ?: return null
        }
    }
}
